File: src/studio/backend/data/params.py
"""
Parameters: flattening of raw experiment JSON, the search grammar, and schedule curves.

Search grammar: `term (AND term)*`. A term is either `path_suffix op value` with
`op` in `=, !=, >, <, >=, <=, ~`, or free text. Whitespace-separated terms are also combined with AND.
"""

import math
import re
from collections.abc import Iterable
from dataclasses import dataclass, field
from typing import Any, Literal

from .extract import CLASS_KEY, as_int

ParamKind = Literal["object", "schedule", "array", "number", "string", "boolean", "null"]
Op = Literal["=", "!=", ">", "<", ">=", "<=", "~"]
CURVE_POINTS = 64


@dataclass
class ParamRow:
    path: str
    key: str
    depth: int
    kind: ParamKind
    value: Any
    cls: str | None = None
    curve: dict[str, list[float]] | None = None

    def to_json(self) -> dict[str, Any]:
        """The API's `ParamRow` (non-finite floats become null). @ai-generated"""
        value = self.value
        if isinstance(value, float) and not math.isfinite(value):
            value = None
        return {
            "path": self.path,
            "key": self.key,
            "depth": self.depth,
            "kind": self.kind,
            "value": value,
            "cls": self.cls,
            "curve": self.curve,
        }


# ---------------------------------------------------------------- Schedules


def _num(node: dict, key: str) -> float | None:
    """@ai-generated"""
    value = node.get(key)
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    return float(value)


def schedule_value(node: Any, t: float) -> float | None:
    """
    Value at time step `t` of a serialized schedule, mirroring `marl.utils.schedule`
    (Linear, Exp, Constant and Rounded schedules). None for unknown or malformed schedules.

    @ai-generated
    """
    if not isinstance(node, dict):
        return None
    cls = node.get(CLASS_KEY)
    if cls == "RoundedSchedule":
        inner = schedule_value(node.get("schedule"), t)
        digits = as_int(node.get("n_digits", 0))
        return None if inner is None or digits is None else round(inner, digits)
    start = _num(node, "start_value")
    if cls == "ConstantSchedule":
        return start
    end, n = _num(node, "end_value"), _num(node, "n_steps")
    if start is None or end is None or n is None:
        return None
    if t >= n:
        return end
    if cls == "LinearSchedule":
        return (end - start) / n * t + start
    if cls == "ExpSchedule":
        if n <= 1 or start == 0:
            return start
        try:
            return start * (end / start) ** (t / (n - 1))
        except (ZeroDivisionError, OverflowError, ValueError):
            return None
    return None


def schedule_curve(node: Any, horizon: int | None = None, points: int = CURVE_POINTS) -> dict[str, list[float]] | None:
    """
    Sampled curve `{x, y}` of a schedule over `[0, max(n_steps, horizon)]`. None if unknown.

    @ai-generated
    """
    if schedule_value(node, 0) is None:
        return None
    n = _schedule_steps(node)
    end = max(n or 0, horizon or 0) or 1
    xs = sorted({round(end * i / (points - 1)) for i in range(points)} | ({n} if n and n <= end else set()))
    ys = [schedule_value(node, x) for x in xs]
    if any(y is None or not math.isfinite(y) for y in ys):
        return None
    return {"x": [float(x) for x in xs], "y": [float(y) for y in ys]}  # type: ignore[arg-type]


def _schedule_steps(node: dict) -> int | None:
    """@ai-generated"""
    if node.get(CLASS_KEY) == "RoundedSchedule":
        inner = node.get("schedule")
        return _schedule_steps(inner) if isinstance(inner, dict) else None
    return as_int(node.get("n_steps"))


# ---------------------------------------------------------------- Flatten


def _is_scalar(value: Any) -> bool:
    return value is None or isinstance(value, (bool, int, float, str))


def _leaf_kind(value: Any) -> ParamKind:
    """@ai-generated"""
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "boolean"
    if isinstance(value, (int, float)):
        return "number"
    return "string"


def flatten(raw: Any, horizon: int | None = None) -> list[ParamRow]:
    """
    Depth-first rows of a raw JSON object. `class-name` becomes `cls` of object nodes, `name` keys
    duplicating the class name are hidden, and lists of scalars are leaves. Schedules (class name
    ending in `Schedule`) get a curve over `horizon` (default: the root's `n_steps`).

    @ai-generated
    """
    if not isinstance(raw, dict):
        return []
    if horizon is None:
        horizon = as_int(raw.get("n_steps"))
    rows = list[ParamRow]()

    def visit_children(node: Any, prefix: str, depth: int):
        if isinstance(node, dict):
            cls = node.get(CLASS_KEY)
            items: Iterable[tuple[str, Any]] = ((k, v) for k, v in node.items() if k != CLASS_KEY and not (k == "name" and v == cls))
        else:
            items = ((str(i), v) for i, v in enumerate(node))
        for key, value in items:
            visit(key, value, f"{prefix}.{key}" if prefix else key, depth)

    def visit(key: str, value: Any, path: str, depth: int):
        if isinstance(value, dict):
            cls = value.get(CLASS_KEY)
            cls = cls if isinstance(cls, str) else None
            if cls is not None and cls.endswith("Schedule"):
                rows.append(ParamRow(path, key, depth, "schedule", None, cls, schedule_curve(value, horizon)))
            else:
                rows.append(ParamRow(path, key, depth, "object", None, cls))
            visit_children(value, path, depth + 1)
        elif isinstance(value, list):
            if all(_is_scalar(v) for v in value):
                rows.append(ParamRow(path, key, depth, "array", value))
            else:
                rows.append(ParamRow(path, key, depth, "array", None))
                visit_children(value, path, depth + 1)
        else:
            rows.append(ParamRow(path, key, depth, _leaf_kind(value), value))

    visit_children(raw, "", 0)
    return rows


# ---------------------------------------------------------------- Search


_TOKEN_RE = re.compile(r'(?P<path>[\w.\-\[\]]+)\s*(?P<op>>=|<=|!=|=|>|<|~)\s*(?P<value>"[^"]*"|\S+)|(?P<text>"[^"]*"|\S+)')


@dataclass(frozen=True)
class Term:
    path: str | None
    """None for free text."""
    op: Op | None
    value: str


@dataclass
class SearchContext:
    """What a query is matched against: identity fields and flattened parameters."""

    id: str
    name: str
    algo: str | None = None
    env: str | None = None
    test_env: str | None = None
    status: str | None = None
    rows: list[ParamRow] = field(default_factory=list)

    def all_rows(self) -> list[ParamRow]:
        """Parameter rows plus virtual rows for the identity fields. @ai-generated"""
        virtual = [
            ParamRow(key, key, 0, "string", value)
            for key, value in (("id", self.id), ("name", self.name), ("algo", self.algo), ("env", self.env), ("status", self.status))
            if value is not None
        ]
        return [*virtual, *self.rows]


def parse_query(q: str) -> list[Term]:
    """@ai-generated"""
    terms = list[Term]()
    for match in _TOKEN_RE.finditer(q or ""):
        if match.group("text") is not None:
            text = match.group("text").strip('"')
            if text and text.upper() != "AND":
                terms.append(Term(None, None, text))
        else:
            terms.append(Term(match.group("path"), match.group("op"), match.group("value").strip('"')))  # type: ignore[arg-type]
    return terms


def _to_number(value: str) -> float | None:
    """@ai-generated"""
    try:
        return float(value)
    except ValueError:
        return None


def _compare_row(row: ParamRow, op: Op, value: str) -> bool:
    """Whether `row <op> value` holds; object and schedule nodes compare their class name. @ai-generated"""
    target: Any = row.cls if row.kind in ("object", "schedule") else row.value
    if row.kind == "array":
        target = row.value
        if op == "~" and isinstance(target, list):
            return any(value.casefold() in str(v).casefold() for v in target)
        return False
    if op == "~":
        return target is not None and value.casefold() in str(target).casefold()
    number = _to_number(value)
    if isinstance(target, (int, float)) and not isinstance(target, bool) and number is not None:
        return {
            "=": target == number,
            "!=": target != number,
            ">": target > number,
            "<": target < number,
            ">=": target >= number,
            "<=": target <= number,
        }[op]
    if op not in ("=", "!="):
        return False
    if target is None:
        equal = value.casefold() in ("null", "none")
    elif isinstance(target, bool):
        equal = str(target).casefold() == value.casefold()
    else:
        equal = str(target).casefold() == value.casefold()
    return equal if op == "=" else not equal


def _path_matches(path: str, suffix: str) -> bool:
    return path == suffix or path.endswith("." + suffix)


def match_term(term: Term, ctx: SearchContext, rows: list[ParamRow] | None = None) -> bool:
    """@ai-generated"""
    rows = ctx.all_rows() if rows is None else rows
    if term.path is None or term.op is None:
        needle = term.value.casefold()
        for row in rows:
            for candidate in (row.cls, row.value if row.kind != "array" else None):
                if candidate is not None and not isinstance(candidate, bool) and needle in str(candidate).casefold():
                    return True
        return False
    candidates = [row for row in rows if _path_matches(row.path, term.path)]
    if not candidates:
        return False
    if term.op == "!=":
        return all(_compare_row(row, "!=", term.value) for row in candidates)
    return any(_compare_row(row, term.op, term.value) for row in candidates)


def matches(query: str | list[Term], ctx: SearchContext) -> bool:
    """All terms of the query match (an empty query matches everything). @ai-generated"""
    terms = parse_query(query) if isinstance(query, str) else query
    rows = ctx.all_rows()
    return all(match_term(term, ctx, rows) for term in terms)

File: src/studio/frontend/src/domain/params.ts
/**
 * Parameters: flatten `experiment.json` into rows, compare experiments, and compute schedule
 * curves. Mirrors the backend semantics (backend.md §7) so both produce the same rows:
 *
 * - `class-name` becomes `cls` on object nodes (it is not a row itself);
 * - a `name` key whose value duplicates the node's class name is hidden;
 * - arrays are leaves (`kind: "array"`), whatever they contain;
 * - nodes whose class ends in `Schedule` have `kind: "schedule"` and, when the class is known,
 *   a `curve` mirroring `marl.utils.schedule`.
 */
import type { Curve, ParamKind, ParamRow } from "../api/schemas";
import type { ParamScalar } from "./colour";

const isPlainObject = (v: unknown): v is Record<string, unknown> => typeof v === "object" && v !== null && !Array.isArray(v);

/** Class name of an object node, if any. */
export function className(v: unknown): string | null {
  return isPlainObject(v) && typeof v["class-name"] === "string" ? v["class-name"] : null;
}

export function isSchedule(v: unknown): boolean {
  return /Schedule$/.test(className(v) ?? "");
}

function kindOf(v: unknown): ParamKind {
  if (v === null || v === undefined) return "null";
  if (Array.isArray(v)) return "array";
  if (isPlainObject(v)) return isSchedule(v) ? "schedule" : "object";
  if (typeof v === "number") return "number";
  if (typeof v === "boolean") return "boolean";
  return "string";
}

/**
 * Flatten a raw parameter tree depth-first into `ParamRow`s.
 *
 * @ai-generated
 */
export function flatten(raw: Record<string, unknown>, prefix = "", depth = 0, out: ParamRow[] = []): ParamRow[] {
  const cls = className(raw);
  for (const [key, v] of Object.entries(raw)) {
    if (key === "class-name") continue;
    if (key === "name" && cls !== null && v === cls) continue;
    const path = prefix ? `${prefix}.${key}` : key;
    const kind = kindOf(v);
    if (kind === "object" || kind === "schedule") {
      const node = v as Record<string, unknown>;
      out.push({ path, key, depth, kind, value: null, cls: className(node), curve: kind === "schedule" ? scheduleCurve(node) : null });
      flatten(node, path, depth + 1, out);
    } else {
      out.push({ path, key, depth, kind, value: v === undefined ? null : v, cls: null, curve: null });
    }
  }
  return out;
}

/**
 * Value of schedule node `s` at time step `t`, or `null` for unknown schedule classes.
 * Mirrors `LinearSchedule`, `ExpSchedule`, `ConstantSchedule` and `RoundedSchedule`.
 *
 * @ai-generated
 */
export function scheduleValue(s: Record<string, unknown>, t: number): number | null {
  const cls = className(s);
  const start = Number(s.start_value);
  const end = Number(s.end_value);
  const n = Number(s.n_steps);
  switch (cls) {
    case "ConstantSchedule":
      return Number.isFinite(start) ? start : null;
    case "LinearSchedule":
      if (![start, end, n].every(Number.isFinite) || n <= 0) return null;
      return t >= n ? end : start + ((end - start) / n) * t;
    case "ExpSchedule":
      if (![start, end, n].every(Number.isFinite) || n <= 1 || start === 0) return null;
      return t >= n ? end : start * (end / start) ** (t / (n - 1));
    case "RoundedSchedule": {
      const inner = s.schedule;
      if (!isPlainObject(inner)) return null;
      const v = scheduleValue(inner, t);
      const digits = Number(s.n_digits ?? 0);
      if (v === null) return null;
      const f = 10 ** (Number.isFinite(digits) ? digits : 0);
      return Math.round(v * f) / f;
    }
    default:
      return null;
  }
}

/** Sample a schedule over [0, 1.25·n_steps] (at least [0, 1]); `null` for unknown classes. @ai-generated */
export function scheduleCurve(s: Record<string, unknown>, points = 40): Curve | null {
  const inner = className(s) === "RoundedSchedule" && isPlainObject(s.schedule) ? s.schedule : s;
  const n = Number(inner.n_steps);
  const total = Math.max((Number.isFinite(n) ? n : 0) * 1.25, 1);
  const x: number[] = [];
  const y: number[] = [];
  for (let i = 0; i <= points; i++) {
    const t = (i / points) * total;
    const v = scheduleValue(s, t);
    if (v === null) return null;
    x.push(t);
    y.push(v);
  }
  return { x, y };
}

/**
 * Comparable value of a row: object/schedule nodes compare by class name, leaves by value
 * (arrays are JSON-encoded so they compare by content).
 *
 * @ai-generated
 */
export function rowValue(row: ParamRow): ParamScalar {
  if (row.kind === "object" || row.kind === "schedule") return row.cls ?? "{…}";
  const v = row.value;
  if (Array.isArray(v)) return JSON.stringify(v);
  if (v === null || typeof v === "string" || typeof v === "number" || typeof v === "boolean") return v;
  return JSON.stringify(v);
}

/** Value of parameter `path` in `rows` (`undefined` when absent). @ai-generated */
export function paramValue(rows: ParamRow[] | undefined, path: string): ParamScalar | undefined {
  const r = rows?.find((x) => x.path === path);
  return r ? rowValue(r) : undefined;
}

/** Human formatting of a parameter value (mockup's `fmtValue`). @ai-generated */
export function formatParamValue(v: unknown): string {
  if (v === null || v === undefined) return "—";
  if (Array.isArray(v)) return `[${v.join(", ")}]`;
  if (typeof v === "number") {
    if (Number.isInteger(v)) return Math.abs(v) >= 10000 ? v.toLocaleString("en-US") : String(v);
    if (v !== 0 && Math.abs(v) < 1e-3) return v.toExponential(1);
    return String(+v.toPrecision(4));
  }
  return String(v);
}

export const ABSENT = "∅";

export type DiffRow = {
  path: string;
  key: string;
  depth: number;
  kind: ParamKind;
  /** Display value per experiment, in the order of the input (`∅` when absent). */
  values: string[];
  rows: (ParamRow | undefined)[];
  differs: boolean;
};

/**
 * Union of parameter paths across experiments (first-seen order), with each experiment's
 * display value and whether they differ.
 *
 * @ai-generated
 */
export function diff(sets: ParamRow[][]): DiffRow[] {
  const maps = sets.map((rows) => new Map(rows.map((r) => [r.path, r])));
  const paths: string[] = [];
  const seen = new Set<string>();
  for (const rows of sets)
    for (const r of rows)
      if (!seen.has(r.path)) {
        seen.add(r.path);
        paths.push(r.path);
      }
  return paths.map((path) => {
    const rows = maps.map((m) => m.get(path));
    const any = rows.find((r): r is ParamRow => r !== undefined)!;
    const values = rows.map((r) => (r ? (r.kind === "object" || r.kind === "schedule" ? (r.cls ?? "{…}") : formatParamValue(r.value)) : ABSENT));
    return { path, key: any.key, depth: any.depth, kind: any.kind, values, rows, differs: new Set(values).size > 1 };
  });
}
